File: Data Structures & Algorithms/subsets/submission-0.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:

        subsets = []

        def dfs(i, cur_subset):
            if i >= len(nums):
                subsets.append(cur_subset[:])
                return
            
            cur_subset.append(nums[i])
            dfs(i + 1, cur_subset)
            cur_subset.pop()
            dfs(i + 1, cur_subset)
        
        dfs(0, [])

        return subsets

File: Data Structures & Algorithms/combinations/submission-0.java
class Solution {
    private void dfs(int[] nums, int i, int k, List<Integer> subset, List<List<Integer>> subsets) {
        if (subset.size() == k) {
            subsets.add(new ArrayList<>(subset));
            return;
        }

        if (i == nums.length) return;

        subset.add(nums[i]);
        dfs(nums, i + 1, k, subset, subsets);
        subset.remove(subset.size() - 1);
        dfs(nums, i + 1, k, subset, subsets);
    }

    public List<List<Integer>> combine(int n, int k) {
        int[] nums = new int[n];

        for (int i = 0; i < n; i++) {
            nums[i] = i + 1;
        }

        List<List<Integer>> subsets = new ArrayList<>();
        dfs(nums, 0, k, new ArrayList<>(), subsets);
        return subsets;
    }
}